File: courses/Natural Language Processing (Stanford 2012)/Assignment/pa2-autocorrect-v1/StupidBackoffLanguageModel.py
import collections, math

class StupidBackoffLanguageModel:

  def __init__(self, corpus):
    """Initialize your data structures in the constructor."""
    # TODO your code here
    self.unigramCounts = collections.defaultdict(lambda: 0)
    self.bigramCounts = collections.defaultdict(lambda: 0)
    self.trigramCounts = collections.defaultdict(lambda: 0)
    self.total = 0
    self.train(corpus)

  def train(self, corpus):
    """ Takes a corpus and trains your language model. 
        Compute any counts or other corpus statistics in this function.
    """  
    # TODO your code here
    for sentence in corpus.corpus:
      prev_token1, prev_token2 = None, None
      for datum in sentence.data:  
        token = datum.word
        self.unigramCounts[token] += 1
        self.total += 1  
        if prev_token2:    
           self.bigramCounts[prev_token2, token] += 1
        if prev_token1:    
           self.trigramCounts[prev_token1, prev_token2, token] += 1
        prev_token1, prev_token2 = prev_token2, token

  def score(self, sentence):
    """ Takes a list of strings as argument and returns the log-probability of the 
        sentence using your language model. Use whatever data you computed in train() here.
    """
    # TODO your code here
    score = 0.0 
    prev_token1, prev_token2 = None, None
    alpha = 0.4 #1
    for token in sentence:
      if prev_token2 == None:
          count = self.unigramCounts[token]
          score += math.log(count + 1)
          score -= math.log(self.total + len(self.unigramCounts))
      elif prev_token1 == None:
          count2 = self.bigramCounts[prev_token2,token]
          count1 = self.unigramCounts[prev_token2]
          if count2 > 0:
			  score += math.log(count2)
			  score -= math.log(count1)
          else:
			  count = self.unigramCounts[token]
			  score += alpha*math.log(count + 1)
			  score -= alpha*math.log(self.total + len(self.unigramCounts))
      else:
          count3 = self.trigramCounts[prev_token1,prev_token2,token]
          count2 = self.bigramCounts[prev_token1,prev_token2]
          if count3 > 0:
			  score += math.log(count3)
			  score -= math.log(count2)
          else:
			  count2 = self.bigramCounts[prev_token2,token]
			  count1 = self.unigramCounts[prev_token2]
			  if count2 > 0:
				  score += alpha*math.log(count2)
				  score -= alpha*math.log(count1)
			  else:
				  count = self.unigramCounts[token]
				  score += alpha*math.log(count + 1)
				  score -= alpha*math.log(self.total + len(self.unigramCounts))
      prev_token1, prev_token2 = prev_token2, token
        
    return score

File: courses/Natural Language Processing (Stanford 2012)/Assignment/pa2-autocorrect-v1/StupidBackoffLanguageModel.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class StupidBackoffLanguageModel implements LanguageModel {

  protected Map<String, Integer> wordFreqMap; // map of words that occur in training with their corresponding frequencies
  protected Map<String, Integer> wordPairFreqMap; // map of word pairs that occur in training with their corresponding frequencies
  int totWordType = 0;	
  int totWordFreq = 0; // total # of words in the training corpus

  /** Initialize your data structures in the constructor. */
  public StupidBackoffLanguageModel(HolbrookCorpus corpus) {
    wordFreqMap = new HashMap<String, Integer>();
    wordPairFreqMap = new HashMap<String, Integer>();
    train(corpus);
  }

  /** Takes a corpus and trains your language model. 
    * Compute any counts or other corpus statistics in this function.
    */
  public void train(HolbrookCorpus corpus) {
	for(Sentence sentence : corpus.getData()) { // iterate over sentences
      String word = null, prevWord = null, wordPair = null;
	  for(Datum datum : sentence) { // iterate over words
        prevWord = word;
		word = datum.getWord(); // get the actual word
		Integer freq = wordFreqMap.get(word);
		int f = freq == null ? 1 : freq.intValue() + 1;
        wordFreqMap.put(word, f);
		if (prevWord != null) {
			wordPair = prevWord + "," + word;
			freq = wordPairFreqMap.get(wordPair);
			f = freq == null ? 1 : freq.intValue() + 1;
			wordPairFreqMap.put(wordPair, f);
		}
      }
    }
	totWordType = wordFreqMap.size();
	for (Map.Entry<String, Integer> entry : wordFreqMap.entrySet()) { // iterate over map
		totWordFreq += entry.getValue().intValue();
	}
  }

  /** Takes a list of strings as argument and returns the log-probability of the 
    * sentence using your language model. Use whatever data you computed in train() here.
    */
  public double score(List<String> sentence) {
    double score = 0.0;
	String prevWord = null;
    int fpair, f, fprev;
	fpair = f = fprev = 0;
	double probability = 0.0;
	for(String word : sentence) { // iterate over words in the sentence
	  Integer freq = wordFreqMap.get(word);
	  f = freq == null ? 0 : freq.intValue();
	  if (prevWord != null) {
		  String wordPair = prevWord + "," + word;
		  Integer freqPair = wordPairFreqMap.get(wordPair), freqPrev = wordFreqMap.get(prevWord);
		  fpair = freqPair == null ? 0 : freqPair.intValue();
		  fprev = freqPrev == null ? 0 : freqPrev.intValue();
		  probability = Math.log(fpair > 0 ? (fpair * 1.0) / fprev : (f + 1.0) / (totWordFreq + totWordType));
	  }
	  else {
	  	  probability = Math.log((f + 1.0) / (totWordFreq + totWordType));
 	  }
	  prevWord = word;
      score += probability;
   }
    return score;
  }
}
